module/mempool/stdmap/backdata/arrayBackData.go

```diff
@@ -0,0 +1,400 @@
+			// as an available slot.
+			availableSlotCount++
+			slotToUse = s
+			continue
+		}
+
+		if id != entityId {
+			// key is fresh, fully distinct, and linked. Hence,
+			// moving to next key.
+			continue
+		}
+
+		// entity ID already exists in the bucket
+		a.availableSlotHistogram[availableSlotCount]++
+		return 0, false
+	}
+
+	a.availableSlotHistogram[availableSlotCount]++
+	return slotToUse, true
+}
+
+// ownerIndexOf maps the (bucketIndex, slotIndex) pair to a canonical unique (scalar) index.
+// This scalar index is used to represent this (bucketIndex, slotIndex) pair in the underlying
+// entities list.
+func (a ArrayBackData) ownerIndexOf(bucketIndex bIndex, slotIndex sIndex) uint64 {
+	return (uint64(bucketIndex) * bucketSize) + uint64(slotIndex)
+}
+
+// linkedEntityOf returns the entity linked to this (bucketIndex, slotIndex) pair from the underlying entities list.
+// By a linked entity, we mean if the entity has an owner index matching to (bucketIndex, slotIndex).
+// The bool return value corresponds to whether there is a linked entity to this (bucketIndex, slotIndex) or not.
+func (a *ArrayBackData) linkedEntityOf(bucketIndex bIndex, slotIndex sIndex) (flow.Identifier, flow.Entity, bool) {
+	if a.buckets[bucketIndex][slotIndex].keyIndex == 0 {
+		// slotIndex never used, or recently invalidated, hence
+		// does not have any linked entity
+		return flow.Identifier{}, nil, false
+	}
+
+	// retrieving entity index in the underlying entities linked-list
+	valueIndex := a.buckets[bucketIndex][slotIndex].entityIndex
+	id, entity, owner := a.entities.Get(valueIndex)
+	if a.ownerIndexOf(bucketIndex, slotIndex) != owner {
+		// entity is not linked to this (bucketIndex, slotIndex)
+		a.buckets[bucketIndex][slotIndex].keyIndex = 0
+		return flow.Identifier{}, nil, false
+	}
+
+	return id, entity, true
+}
+
+// logTelemetry prints telemetry logs depending on number of interactions and
+// last time telemetry has been logged.
+func (a *ArrayBackData) logTelemetry() {
+	a.interactionCounter++
+	if a.interactionCounter < telemetryCounterInterval {
+		// not enough interactions to log.
+		return
+	}
+	if time.Duration(runtimeNano()-a.lastTelemetryDump) < telemetryDurationInterval {
+		// not long elapsed since last log.
+		return
+	}
+
+	lg := a.logger.With().
+		Uint64("total_keys_written", a.keyCount).
+		Uint64("total_interactions_since_last_log", a.interactionCounter).Logger()
+
+	for i := range a.availableSlotHistogram {
+		lg = lg.With().
+			Int("available_slots", i).
+			Uint64("total_buckets", a.availableSlotHistogram[i]).
+			Logger()
+	}
+
+	lg.Debug().Msg("logging telemetry")
+	a.interactionCounter = 0
+	a.lastTelemetryDump = runtimeNano()
+}
+
+// invalidateKey sets the key index of specified slot in the bucket to zero, so the key
+// is free to take.
+func (a *ArrayBackData) invalidateKey(bucketIndex bIndex, slotIndex sIndex) {
+	a.buckets[bucketIndex][slotIndex].keyIndex = 0
```

Comment:
Consider making `0` into a const e.g. ` = slotAgeUnallocated` ?

If `keyIndex` was refactored to `slotAge` then we'd get `.slotAge = slotAgeUnallocated` :-)

Reply:
Applied in https://github.com/onflow/flow-go/pull/1778/commits/e4b057bbb9fc407bad53c664b660431442882b9f